test/validation/api/ipsec/ipsec.c

```diff
@@ -878,6 +878,11 @@ static void verify_in(const ipsec_test_part *part,
 			if (ODP_IPSEC_SA_INVALID != sa)
 				CU_ASSERT_EQUAL(IPSEC_SA_CTX,
 						odp_ipsec_sa_context(sa));
+			if (suite_context.inbound_op_mode != ODP_IPSEC_OP_MODE_SYNC) {
+				uint32_t len = part->pkt_in->len - part->pkt_in->l3_offset;
+
+				CU_ASSERT(result.orig_ip_len == len);
```

Comment:
Wouldn't this ASSERT make result.orig_ip_len mandatory?

API text makes it optional.

> Zero value means that the length information is not available.

Reply:
@JannePeltonen There is one more issue here. In case of reassembly test cases, part->pkt_in would be just one fragment (in fact last fragment). So this would be comparing IP packet length of one fragment against IP packet length of all fragments, which is not the intention,

There is no quick way of fixing this. I have added a fix, but that would involve a new member ``ipsec_test_part.orig_ip_len`` and I would populate this field in case of reassembly tests.

And the change in verify_in would be like,
`+                               uint32_t len;                                                                                                                                                                                                                                                                                                                                           +                               if (part->orig_ip_len)                                                                                                                                                             +                                       len = part->orig_ip_len;                                                                                                                                                   +                               else                                                                                                                                                                               +                                       len = part->pkt_in->len - part->pkt_in->l3_offset; `

Another option is to set part->orig_ip_len in all cases, which would lead to lot of code duplication.

I can push a patch if you agree that we have a problem and you agree with either solution.

Apologies that I didn't review this patch earlier.

Reply:
> The new orig_ip_len field of test_part could have some comment that says that it is nonzero only if the expected len differs from that of the input test packet.

Agreed. And as I worked on it further, it looked like making orig_ip_len as part of ipsec_test_part.out[i] seemed like a better approach. For incomplete frags cases, there can be more than 1 frag and single value won't work for all frags. Shall I propose so?

> I would think that the value of the expected orig_ip_len could be calculated on the fly in test_multi_out_in() from the test fragments so that it would not need to be hardcoded in the reassembly test cases.

That is the intention. We are on the same page.

Reply:
> it looked like making orig_ip_len as part of ipsec_test_part.out[i] seemed like a better approach

That sounds right, even though the test code does not appear to be ready to handle more than one fragment in the incomplete case (either that or I got confused reading the code). But that is another issue and not a problem until such a test case gets added.

Reply:
Agreed. I'll push a PR with the discussed changes.